pkg/resource/config/map.go

```diff
@@ -47,6 +55,253 @@ func (m Map) HasSecureValue() bool {
+	if len(p) == 1 {
+		delete(m, configKey)
+		return nil
+	}
+
+	// Otherwise, lookup the current root value and save it into a temporary map.
+	root := make(map[string]interface{})
+	if val, ok := m[configKey]; ok {
+		obj, err := val.ToObject()
+		if err != nil {
+			return err
+		}
+		root[configKey.Name()] = obj
+	}
+
+	// Get the value within the object up to the second-to-last path segment.
+	// If not found, exit early.
+	parent, dest, ok := getValueForPath(root, p[:len(p)-1])
+	if !ok {
+		return nil
+	}
+
+	// Remove the last path segment.
+	key := p[len(p)-1]
+	switch t := dest.(type) {
+	case []interface{}:
+		index, ok := key.(int)
+		if !ok || index < 0 || index >= len(t) {
+			return nil
+		}
+		t = append(t[:index], t[index+1:]...)
+		// Since we changed the array, we need to update the parent.
+		if parent != nil {
+			pkey := p[len(p)-2]
+			if _, err := setValue(parent, pkey, t, nil, nil); err != nil {
+				return err
+			}
+		}
+	case map[string]interface{}:
+		k, ok := key.(string)
+		if !ok {
+			return nil
+		}
+		delete(t, k)
+
+		// Secure values are reserved, so return an error when attempting to add one.
+		if isSecure, _ := isSecureValue(t); isSecure {
+			return errSecureKeyReserved
+		}
+	}
+
+	// Now, marshal then unmarshal the value, which will handle detecting
+	// whether it's a secure object or not.
+	jsonBytes, err := json.Marshal(root[configKey.Name()])
+	if err != nil {
+		return err
+	}
+	var v Value
+	if err = json.Unmarshal(jsonBytes, &v); err != nil {
+		return err
+	}
+
+	m[configKey] = v
+	return nil
+}
+
+// Set sets the value for a given key. If path is true, the key's name portion is treated as a path.
+func (m Map) Set(k Key, v Value, path bool) error {
+	// If the key isn't a path, go ahead and set the value and return.
+	if !path {
+		m[k] = v
+		return nil
+	}
+
+	// Otherwise, parse the path and get the new config key.
+	p, configKey, err := parseKeyPath(k)
+	if err != nil {
+		return err
+	}
+
+	// If we only have a single path segment, set the value and return.
+	if len(p) == 1 {
+		m[configKey] = v
+		return nil
+	}
+
+	// Otherwise, lookup the current value and save it into a temporary map.
+	root := make(map[string]interface{})
+	if val, ok := m[configKey]; ok {
+		obj, err := val.ToObject()
+		if err != nil {
+			return err
+		}
+		root[configKey.Name()] = obj
+	}
+
+	// Now, loop through the path segments, and walk the object tree.
+	// If the value for a given segment is nil or isn't the expected type, create a new array/map.
+	// The root map is the initial cursor value, and parent is nil.
+	var parent interface{}
+	var parentKey interface{}
+	var cursor interface{}
+	var cursorKey interface{}
+	cursor = root
+	cursorKey = p[0]
+	for _, pkey := range p[1:] {
+		pvalue, err := getValue(cursor, cursorKey)
+		if err != nil {
+			return err
+		}
+
+		// If the value is nil or isn't the expected type for the path segment key, create a new array/map.
```

Comment:
I think this is okay for `nil`, but is this the behavior we want for type mismatches? Seems like we might want to error/warn in this situation.

Reply:
Changed the behavior to error on type mismatches.